src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -56,75 +58,83 @@ internal class RenamerProjectTreeActionHandler : ProjectTreeActionHandlerBase, I
             _vsOnlineServices = vsOnlineServices;
             _threadingService = threadingService;
             _extensibility = extensibility;
+            _languageServiceDone = new Dictionary<(string, string), SemaphoreSlim>();
         }
 
         public void HandleRename(string oldFilePath, string newFilePath)
         {
-            if (_oldFilePath == oldFilePath)
+            string newFileWithExtension = Path.GetFileName(newFilePath);
+            if (_languageServiceDone.TryGetValue((oldFilePath, newFileWithExtension), out SemaphoreSlim semLanguegeServiceDone))
```

Comment:
If this is called before RenameAsync adds to the list, RenameAsync will wait forever.

Reply:
My understanding is that we should wait for `HandleRename `to happen. In this case, a semaphore with release status is added when before `RenameAsyn`.